### DISCUSSION TITLE
慢查询问题

### DISCUSSION THREAD

**Discussion by yaobo-lab** [dotnetcore/FreeSql#994] [Q&A]
```
fsql.Aop.CurdAfter += (s, e) =>
          {                    
              if (e.ElapsedMilliseconds >=2000)
              {                    
                  Console.WriteLine($"\n Sql 慢查询: \n {e.Sql}, \n 耗时:{e.ElapsedMilliseconds} 毫秒");
              }
          };
```

请问我这边记录超两秒的SQL语句，但具体是那个方法生成的SQL 导致慢查询。我是不清楚的。
我查了下委托方法的参数好像也没有见到。

请问我如何知道是那个方法。生成的SQL语句是慢查询？

**Comment by luoyunchong** [comment]
`CurdAfterEventArgs` 继承 `CurdBeforeEventArgs`,目前只能得到访问的哪个表，无法得到具体是什么方法。

```cs
        /// <summary>
        /// 实体类型
        /// </summary>
        public Type EntityType { get; }
        /// <summary>
        /// 实体类型的元数据
        /// </summary>
        public TableInfo Table { get; }
```

**Reply by yaobo-lab** [reply]
> `CurdAfterEventArgs` 继承 `CurdBeforeEventArgs`,目前只能得到访问的哪个表，无法得到具体是什么方法。
> 
> ```cs
>         /// <summary>
>         /// 实体类型
>         /// </summary>
>         public Type EntityType { get; }
>         /// <summary>
>         /// 实体类型的元数据
>         /// </summary>
>         public TableInfo Table { get; }
> ```

那没有意义呢。 业务逻辑多变，不能保证是那个程序员写业务代码不合格。导致 生成的SQL 没有使用到索引。 只打印SQL 基本很难定位问题所在的方法

**Reply by luoyunchong** [reply]
另外做所有请求方法的日志吧。

**Reply by yaobo-lab** [reply]
> 所有请求方法的日志吧。

请求日志又是不一样的性质了。请求耗时太长这个原因不单单是db的问题

**Reply by yaobo-lab** [reply]
> `CurdAfterEventArgs` 继承 `CurdBeforeEventArgs`,目前只能得到访问的哪个表，无法得到具体是什么方法。
> 
> ```cs
>         /// <summary>
>         /// 实体类型
>         /// </summary>
>         public Type EntityType { get; }
>         /// <summary>
>         /// 实体类型的元数据
>         /// </summary>
>         public TableInfo Table { get; }
> ```

提供的这些参数意义不大。因为fsql 已经打印了慢查询的语句。当然也就知道是那些表了

**Comment by hd2y** [comment]
可以尝试在日志中打印 `Environment.StackTrace`。

**Reply by yaobo-lab** [reply]
> 可以尝试在日志中打印 `Environment.StackTrace`。

不行。

**Reply by hd2y** [reply]
是不能打印出来堆栈跟踪信息，还是打印出来的信息，无法确认是哪个方法？  
这个属于调试问题，如果这个方案不能满足，只能手动埋点进行测试了。  
可以借助 `Fody` 的 `Tracer` 或者 `MethodTimer` 插件对怀疑可能出现问题的方法进行埋点：https://github.com/Fody/Home/blob/master/pages/addins.md
